src/CoreWCF.Primitives/src/CoreWCF/Security/LdapAdapter.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.DirectoryServices.Protocols;
+using System.Linq;
+using System.Text;
+using CoreWCF.IdentityModel.Claims;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Logging;
+
+namespace CoreWCF.Security
+{
+    internal static class LdapAdapter
+    {
+        public static List<Claim> RetrieveClaims(LdapSettings settings, string userName)
+        {
+            var user = userName;
+            var userAccountNameIndex = user.IndexOf('@');
+            var userAccountName = "";
+            if (userAccountNameIndex == -1)
+            {
+                userAccountNameIndex = user.IndexOf("\\");
+                if (userAccountNameIndex == -1)
+                    return null;
+                userAccountName = user.Substring(userAccountNameIndex+1);
+            }
+            else
+            {
+                userAccountName = user.Substring(0, userAccountNameIndex);
+            }
+
+            List<Claim> roleClaims = new List<IdentityModel.Claims.Claim>();
+            if (settings.ClaimsCache == null)
+            {
+                settings.ClaimsCache = new MemoryCache(new MemoryCacheOptions { SizeLimit = settings.ClaimsCacheSize });
+            }
+
+            if (settings.ClaimsCache.TryGetValue<IEnumerable<string>>(user, out var cachedClaims))
+            {
+                foreach (var claim in cachedClaims)
+                {
+                    roleClaims.Add(new Claim(ClaimTypes.Role, claim, Rights.Identity));
+                }
```

Comment:
Is ClaimTypes.Role the type of claim that gets populated on Windows? I think you probably want the cache to be for `IEnumerable<Claim>` as you want the claimset to include things like the Upn is the ldap record has a userPrincipalName attribute, and a Name claim to hold the sAMAccountName attribute.

Reply:
Thinking about this a little more, instead of `IEnumerable<Claim>`, it should be a `ClaimSet`.

Reply:
If you are asking in aspnet core, yes https://github.com/dotnet/aspnetcore/blob/52eff90fbcfca39b7eb58baad597df6a99a542b0/src/Identity/Extensions.Core/src/ClaimsIdentityOptions.cs#L16 .. but not in WCF... WCF handles via CAS.. while aspnetcore handles via CalimsIdentity for both os